server/server/security.py

```diff
@@ -0,0 +1,70 @@
+"""Implement authentication and authorization policy"""
+from passlib.totp import generate_secret
+from pyramid.authentication import CallbackAuthenticationPolicy
+from pyramid.authorization import ACLAuthorizationPolicy
+from pyramid.interfaces import IAuthenticationPolicy
+from pyramid.security import Authenticated, Everyone
+from zope.interface import implementer
+
+from .models.user import User
+from .models.token import Token
+
+
+@implementer(IAuthenticationPolicy)
+class MyAuthenticationPolicy(CallbackAuthenticationPolicy):
+
+    def unauthenticated_userid(self, request):
+        """Returns user's id by token"""
+        if 'Authorization' in request.headers:
+            token = request.headers['Authorization'].split(' ')[1]
+            token_obj = Token.get_token_obj(request, token)
+            if token_obj is not None:
+                token_obj.update()
+                return token_obj.user_id
+
+    def authenticated_userid(self, request):
+        """Get authenticated user id from user object in request object"""
+        user = request.user
+        if user is not None:
+            return user.id
+
+    def effective_principals(self, request):
+        """Return principals for authenticated user"""
+        principals = [Everyone]
+        user = request.user
+        if user is not None:
+            principals.append(Authenticated)
+            principals.append(str(user.id))
+            principals.append('role:' + user.get_role())
+        return principals
+
+    def remember(self, request, userid, **kw):
+        """Generate random authorization token, save it into db and return"""
+        key = generate_secret()
+        Token.add_token(request, key, userid)
+        return key
+
+    def forget(self, request):
+        """Delete token from db if it exist"""
+        if 'Authorization' in request.headers:
+            if Token.deactivate(request,
+                                request.headers['Authorization'].split(' ')[1]
```

Comment:
You use the same snippet at least twice, so probably you may extract to the function, which returns token extarcted from request.